numpy/_core/src/multiarray/compiled_base.c

```diff
@@ -177,6 +177,11 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
     }
 
     if (lst == NULL) {
+        PyArray_Descr* local_dtype = PyArray_DescrFromType(NPY_INTP);
+        list = PyArray_FromAny(list, local_dtype, 0, 0, NPY_ARRAY_FORCECAST, NULL);
+        if (list == NULL) {
+            goto fail;
+        }
```

Comment:
Neither of the two CI failures look related and I checked that the new regression test indeed fails before/passes after the patch.

I suspect Nathan or Sebastian will have to babysit me on the reference counting/setting for `local_dtype` and/or `list` above perhaps. Anyway, happy to make whatever revisions are requested.

For coercing to `NPY_INTP` we also have `coerce_index` in `iterators.c`, but dropping that in seemed to potentially require a bit more work.

Reply:
Thanks, there are problems here, but the refcounting is fine :).

1. We shouldn't force-cast everything.  We can get away with force-casting (with some pragmatism) only for integer arrays.  So basically, the logic should be: `PyArray_ISINTEGER(list) ? unsafe-cast : safe-cast` (or similar)
2. `PyArray_FromAny` needs the flags to be passed because the later code relies on them.  I.e. we do need the "Coniguous" part here.  That is, it uses `NPY_ARRAY_DEFAULT` (it is just a one line macro for calling `PyArray_FromAny`).
   (Which after a few unpacking, translates to: contiguous, aligned and writeable.)

The current version should crash hard if you do something like `np.bincount(np.arange(10000)[::2])` and it is a bit scary that we have no test for that!

Reply:
(Ah, I spoke too fast, the refcounting is indeed not fine, as user input `list` is a borrowed reference normally and you are assigning a strong one and you need to `DECREF` it.  But once you refactor things, that problem probably goes away by itself.)